rust/core-lib/src/recorder.rs

```diff
@@ -0,0 +1,435 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use xi_trace::trace_block;
+
+use std::mem;
+use std::collections::HashMap;
+
+use edit_types::{BufferEvent, EventDomain};
+
+/// A container that manages and holds all recordings for the current editing session
+pub(crate) struct Recorder {
+    active_recording: Option<String>,
+    recording_buffer: Vec<EventDomain>,
+    recordings: HashMap<String, Recording>,
+}
+
+impl Recorder {
+    pub(crate) fn new() -> Recorder {
+        Recorder {
+            active_recording: None,
+            recording_buffer: Vec::new(),
+            recordings: HashMap::new(),
+        }
+    }
+
+    pub(crate) fn is_recording(&self) -> bool {
+        self.active_recording.is_some()
+    }
+
+    /// Starts or stops the specified recording.
+    ///
+    ///
+    /// There are three outcome behaviors:
+    /// - If the current recording name is specified, the active recording is saved
+    /// - If no recording name is specified, the currently active recording is saved
+    /// - If a recording name other than the active recording is specified,
+    /// the current recording will be thrown out and will be switched to the new name
+    ///
+    /// In addition to the above:
+    /// - If the recording was saved, there is no active recording
+    /// - If the recording was switched, there will be a new active recording
+    pub(crate) fn toggle_recording(&mut self, recording_name: Option<String>) {
+        let is_recording = self.is_recording();
+        let last_recording = self.active_recording.take();
+
+        match (is_recording, &last_recording, &recording_name) {
+            (true, Some(last_recording), None) => self.save_recording_buffer(last_recording.clone()),
+            (true, Some(last_recording), Some(recording_name)) => {
+                if last_recording != recording_name {
+                    self.recording_buffer.clear();
+                } else {
+                    self.save_recording_buffer(last_recording.clone());
+                    return;
+                }
+            },
+            _ => {}
+        }
+
+        mem::replace(&mut self.active_recording, recording_name);
+    }
+
+    /// Saves an event into the currently active recording.
+    /// If no recording is active, the event passed in is ignored.
+    pub(crate) fn record(&mut self, cmd: EventDomain) {
+        if !self.is_recording() {
+            warn!("Recorder not active-- ignoring event {:?}", cmd);
+            return;
+        }
+
+        self.recording_buffer.push(cmd);
```

Comment:
So I've noticed one thing while playing around with this:

![screen shot 2018-10-15 at 6 43 48 pm](https://user-images.githubusercontent.com/3330916/46982456-470f4180-d0aa-11e8-89f1-afb352d9dbcc.png)

This is a trace of playing back a recording that involves numerous `insert` operations. Something that stands out to me are the toothy looking things at the bottom. Each one of those, it turns out, is a call to `Editor::commit_delta`, which is necessary after each buffer event.

This suggests an easy optimization; we can combine any number of sequential inserts into a single insert, which means we'd only have to commit the delta once; this is probably a big enough win to be worth it.

Reply:
I've done as you suggested and merged the inserts when saving the recording. Only annoying thing is that I end up having to `collect` three times.

I took a trace with a test recording that just added `\n\tpub(crate) foo: bool,`.

Before:
<img width="1334" alt="screen shot 2018-10-15 at 8 43 25 pm" src="https://user-images.githubusercontent.com/155693/46991549-4af78f80-d0bb-11e8-9315-e1fde26a8df2.png">

After:
<img width="638" alt="screen shot 2018-10-15 at 8 55 49 pm" src="https://user-images.githubusercontent.com/155693/46991928-bdb53a80-d0bc-11e8-8385-9c1cdea932a4.png">

Only thing I wasn't sure of was whether or not we should merge `InsertNewLine` and `InsertTab` into `Insert` as well; we might start stepping into weird undo territory again since all three of the insert types technically belong to a separate undo group. Maybe merge like inserts only?